### ISSUE TITLE
Performance, in comparison with original AnimateDiff

### ISSUE THREAD

**Issue by actionless** [neggles/animatediff-cli#21]
i'm using same settings (such as number of steps, frames, resolution, scheduler) and both running on cuda.

i've tried 2 different gpus, gtx 1080 and tesla p40

and always i'm getting about 2-3 times slower result with your code (although animation also looks better)

did you noticed the same difference in performance? if not, mb some ideas what i could be overlooking in the settings

also in your code amount of vram used is lower - so i suspect that some computation which in original AnimateDiff done on gpu here are done on cpu (i quickly looked into the code of the pipeline and i saw that for long video `sequential_mode` is used which works on cpu - but i've been testing on 8-16 frames animations, so that shouldn't be the case)

**Comment by actionless** [comment]
```
for config in config/prompts/09-ACTNLZZ.json config/prompts/10-ACTNLZZ.json ; time animatediff generate -c $config -W 512 -H 288 -L 8 -C 8 ; end

Using scheduler "ddim" (DDIMScheduler)
________________________________________________________
Executed in  195.33 secs    fish           external

Using scheduler "k_dpmpp_2m_sde" (DPMSolverMultistepScheduler)
________________________________________________________
Executed in  195.52 secs    fish           external

about 5.5 gb vram

--------------------------------------------------------------------------------

for config in configs/prompts/13-ACTNLZZ.yaml configs/prompts/14-ACTNLZZ.yaml ; time PYTORCH_CUDA_ALLOC_CONF=garbage_collection_threshold:0.6,max_split_size_mb:
21 OMP_NUM_THREADS=12 accelerate launch ./scripts/animate.py --inference_config ./configs/inference/inference.yaml --config $config --L 8 --W 512 --H 288 ; end
________________________________________________________
Executed in   69.63 secs    fish           external
________________________________________________________
Executed in   74.27 secs    fish           external

about 13 gb vram
```

**Comment by actionless** [comment]
oh, i found the problem - i didn't had xformers enabled, here are some additional benchmarks for people interested in fine-tuning the performance of animate diff inference:

```
for config in config/prompts/09-ACTNLZZ.json config/prompts/10-ACTNLZZ.json ; time animatediff generate -c $config -W 512 -H 288 -L 8 -C 8 ; end

Using scheduler "ddim" (DDIMScheduler)
________________________________________________________
Executed in  195.33 secs    fish           external
   usr time  159.45 secs    0.00 micros  159.45 secs
   sys time   52.58 secs  397.00 micros   52.58 secs

Using scheduler "k_dpmpp_2m_sde" (DPMSolverMultistepScheduler)
________________________________________________________
Executed in  195.52 secs    fish           external
   usr time  159.45 secs    0.00 micros  159.45 secs
   sys time   52.71 secs  337.00 micros   52.71 secs

about 5.5 gb vram

--------------------------------------------------------------------------------
accelerate: no difference

+xformers:

Using scheduler "ddim" (DDIMScheduler)
________________________________________________________
Executed in   59.94 secs    fish           external
   usr time   57.77 secs    0.00 micros   57.77 secs
   sys time   16.97 secs  404.00 micros   16.97 secs

Using scheduler "k_dpmpp_2m_sde" (DPMSolverMultistepScheduler)
________________________________________________________
Executed in   60.44 secs    fish           external
   usr time   58.34 secs    0.00 micros   58.34 secs
   sys time   16.99 secs  327.00 micros   16.99 secs

about 5.25 gb vram
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

for config in configs/prompts/14-ACTNLZZ.yaml ; time PYTORCH_CUDA_ALLOC_CONF=garbage_collection_threshold:0.6,max_split_size_mb:
21 OMP_NUM_THREADS=12 accelerate launch ./scripts/animate.py --inference_config ./configs/inference/inference.yaml --config ./configs/prompts/13-ACTNLZZ.yaml --L 8 --W 512 --H 288 ; end

+xformers +accelerate
________________________________________________________
Executed in   69.63 secs    fish           external
   usr time   50.28 secs  482.00 micros   50.28 secs
   sys time   18.20 secs    0.00 micros   18.20 secs

________________________________________________________
Executed in   74.27 secs    fish           external
   usr time   53.83 secs  319.00 micros   53.83 secs
   sys time   19.57 secs  284.00 micros   19.57 secs


about 13 gb vram

--------------------------------------------------------------------------------

no xformers +accelerate:

18 gb vram

Executed in   73.79 secs    fish           external
   usr time   54.11 secs  472.00 micros   54.11 secs
   sys time   19.97 secs    0.00 micros   19.97 secs
--------------------------------------------------------------------------------

no xformers and no accelerate too:

Executed in   73.90 secs    fish           external
   usr time   56.25 secs  213.00 micros   56.25 secs
   sys time   20.03 secs  191.00 micros   20.03 secs

14 gb vram
--------------------------------------------------------------------------------

xformers but no accelerate:

Executed in   66.82 secs    fish           external
   usr time   51.41 secs  190.00 micros   51.41 secs
   sys time   17.85 secs  172.00 micros   17.85 secs

8.5 gb vram
```

benchmark was done on the single tesla, here is the config: 
```
compute_environment: LOCAL_MACHINE
distributed_type: 'NO'
downcast_bf16: 'no'
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

i'm curious if there is a way to somehow spread the load in long video into multiple gpus

**Comment by neggles** [comment]
Hm, that's odd. You shouldn't need xformers to achieve similar levels of performance (since torch's scaled-dot-product-attention impl is generally within a percent or two of xformers' performance); and accelerate is highly unlikely to make a significant difference here since it's primarily focused on distributed processing (and i already have a TorchDynamo model compilation option in the config, which is the other thing accelerate is useful for)

Accelerate *does* appear to be putting all the input latents on GPU, though, rather than leaving them on CPU until they're needed, which is why it consumes so much more VRAM. I have done A/B tests comparing leaving the latents on CPU vs leaving them on GPU and the performance difference is minimal at best, so I've just defaulted to CPU storage because of the massive reduction in VRAM requirement. There's some optimization that could be done there, prefetching latents and such, but it's a bunch of fiddly work for no major benefit 🤷 

as for spreading across multiple GPUs, not really; you could generate multiple animations in parallel, but there's not a practical/performant way to spread a unet across multiple GPUs. While it can be done, the overhead from synchronizing/transferring data between the GPUs results in things being slower than just running on a single GPU :(

**Comment by actionless** [comment]
regarding first 2 paragraphs i have nothing to add - the raw numbers say more than any words or sentences.

also i forgot to mention that in case of using xformers the maximum resolution is limited to BELOW 512x512 even on a server with 24GB gpu-s   - as xformers is utilizing gpu's shared memory for stuff - and performance-wise your code is usable only with xformers - which means it's kinda limited to that small resolution, while OG animatediff performing almost the same with and without xformers.

> as for spreading across multiple GPUs, not really; you could generate multiple animations in parallel, but there's not a practical/performant way to spread a unet across multiple GPUs. While it can be done, the overhead from synchronizing/transferring data between the GPUs results in things being slower than just running on a single GPU :(

LOL - 1) you splitting L-length video to several C-length videos. 2) when seeded with same random - unet running on two different machines would still produce sameish result. -- which means you don't really need to sync anything much, except for once, when need to generate overlap zone between (L/2-O/2 and L/2+O/2 frames) . so you could talk about overhead only in one corner case, where number of gpus is close to `L / C`, e.g. L=48, C=24, num_gpu=2

you probably read in some thread about multi-gpu __learning__ about gradient accumulation overhead, and now just translating that unrelated information to sound like you know things :))

tl;dr i was hoping you decided to refactor animatediff because you know how to make things better, but in fact you [just wanted to ai-hype by poorly re-implementing stuff which wasn't broken at first place](https://en.wikipedia.org/wiki/NIH_syndrome), not cool - i'm very annoyed that spent time figuring out the codebase instead of putting all that energy into original code - considering that i've already added multi-gpu support to OG animatediff earlier, it would be faster to add there sequential animation support, rather than fixing everything here. it just kinda funny that you put effort into coloring output and changing config file format, but never even tried to measure the performance in compare to original 🥲

sorry for overall negative tone of the message - i'm just being triggered when people come up with inaccurate information instead of honestly answering that they dunno smth

**Comment by neggles** [comment]
> regarding first 2 paragraphs i have nothing to add - the raw numbers say more than any words or sentences.
> 
> also i forgot to mention that in case of using xformers the maximum resolution is limited to BELOW 512x512 even on a server with 24GB gpu-s - as xformers is utilizing gpu's shared memory for stuff - and performance-wise your code is usable only with xformers - which means it's kinda limited to that small resolution, while OG animatediff performing almost the same with and without xformers.

I run without xformers, using torch 2.0.1+cu118 or torch 2.1.0 nightly cu121, and at 640x960 it consumes 8.9GB of VRAM:

![qj6Yx35yr7](https://github.com/neggles/animatediff-cli/assets/4232981/3cb94e5b-ffb1-46ed-a57f-30f973f707f2)

I don't even have xformers installed in this venv.

> LOL - 1) you splitting L-length video to several C-length videos. 2) when seeded with same random - unet running on two different machines would still produce sameish result

1. Splitting an L-length video into several C-length videos, while *sort of* doable using some i2i-style implementations in forks of this, isn't the same thing as splitting a *single* generation process using a *single* unet across multiple GPUs, and will not produce the same output as running a single L-length generation on a single GPU.

2. And instantiating two copies of the same Unet with the same random seed on two separate GPUs would just give you two instances of the same output, which is hardly useful.

> you probably read in some thread about multi-gpu **learning** about gradient accumulation overhead, and now just translating that unrelated information to sound like you know things :))

Uh, no. UNets have quite a bit of data crossflow between their outermost layers - that's why they're called **U**-Nets, because they're U-shaped (I wish that was a joke) - which makes splitting a *single* unet model across multiple GPUs come with a *lot* of overhead.

> tl;dr i was hoping you decided to refactor animatediff because you know how to make things better, but in fact you [just wanted to ai-hype by poorly re-implementing stuff which wasn't broken at first place](https://en.wikipedia.org/wiki/NIH_syndrome), not cool - i'm very annoyed that spent time figuring out the codebase instead of putting all that energy into original code - considering that i've already added multi-gpu support to OG animatediff earlier, it would be faster to add there sequential animation support, rather than fixing everything here. it just kinda funny that you put effort into coloring output and changing config file format, but never even tried to measure the performance in compare to original 🥲

I'm not going to dignify this with a response. I spent quite a significant bit of time measuring performance against the original code and my code outstrips it considerably, both in my own tests using xformers and using torch-native SDPA, as well as in tests done by several other users.

Your inability to correctly configure your environment is not my problem.
